File: python/leetcode/876. Middle of the Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        late = head
        early = head.next

        while early:
            early = early.next.next if early.next else None
            late = late.next

        return late

File: go/leetcode/876. Middle of the Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func middleNode(head *ListNode) *ListNode {
    late := head
    early := head.Next
    
    for early != nil {
        early = early.Next
        if early != nil {
            early = early.Next
        } else {early = nil}
        late = late.Next
    }
    
    
    return late
}